In [69]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [70]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [71]:
# import excel file StateGeneralbyPrecinctNov2016.xlsx
tn_2016=pd.read_excel('../data/2016_election/StateGeneralbyPrecinctNov2016.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','RNAME1','PARTY1','PVTALLY1',
                             'RNAME2','PARTY2','PVTALLY2','RNAME3','PARTY3','PVTALLY3','RNAME4','PARTY4','PVTALLY4',
                             'RNAME5','PARTY5','PVTALLY5','RNAME6','PARTY6','PVTALLY6','RNAME7','PARTY7','PVTALLY7',
                             'RNAME8','PARTY8','PVTALLY8','RNAME9','PARTY9','PVTALLY9','RNAME10','PARTY10','PVTALLY10'])
tn_2016

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson     Briceville   
...        ...            ...   
9397    Wilson       Absentee   
9398    Wilson       Absentee   
9399    Wilson       Absentee   
9400    Wilson       Absentee   
9401    Wilson       Absentee   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2016-11-08   
1                               United States President  2016-11-08   
2     United States House of Representatives District 3  2016-11-08   
3        Tennessee House of Representatives District 36  2016-11-08   
4                               United States President  2016-11-08   
...                                                 ...         ...   
9397                            United States President  2016-11-08   
9398                            United States President  2016-11-08   
9399  United States House of Representatives District 6  2016-11-08   
9400     Tennessee House of Representatives District 46  2016-11-08   
9401     Tennessee House of Representatives District 57  2016-11-08   

                              RNAME1       PARTY1  PVTALLY1  \
0                    Donald J. Trump   Republican      1222   
1     Write-In - Kyle Kenley Kopitke          NaN         0   
2                  Chuck Fleischmann   Republican      1124   
3                   Dennis H. Powers   Republican      1128   
4                    Donald J. Trump   Republican       243   
...                              ...          ...       ...   
9397                 Donald J. Trump   Republican       661   
9398  Write-In - Kyle Kenley Kopitke  Independent         0   
9399                     Diane Black   Republican       699   
9400                       Mark Pody   Republican       264   
9401                    Susan M Lynn   Republican       432   

                             RNAME2       PARTY2  PVTALLY2  \
0                   Hillary Clinton   Democratic     227.0   
1     Write-In - Laurence Kotlikoff          NaN       0.0   
2                    Melody Shekari   Democratic     195.0   
3                        Bob Fannon   Democratic     265.0   
4                   Hillary Clinton   Democratic      42.0   
...                             ...          ...       ...   
9397                Hillary Clinton   Democratic     397.0   
9398  Write-In - Laurence Kotlikoff  Independent       0.0   
9399                  David W. Kent   Democratic     297.0   
9400          Amelia Morrison Hipps   Democratic     142.0   
9401                  Trisha Farmer   Democratic     219.0   

                          RNAME3       PARTY3  PVTALLY3  \
0     "Rocky" Roque De La Fuente  Independent       2.0   
1      Write-In - David Limbaugh          NaN       0.0   
2                Topher Kersting  Independent      21.0   
3                            NaN          NaN       NaN   
4     "Rocky" Roque De La Fuente  Independent       3.0   
...                          ...          ...       ...   
9397  "Rocky" Roque De La Fuente  Independent       4.0   
9398   Write-In - David Limbaugh  Independent       0.0   
9399                  David Ross  Independent      71.0   
9400                         NaN          NaN       NaN   
9401                         NaN          NaN       NaN   

                        RNAME4       PARTY4  PVTALLY4  \
0                 Gary Johnson  Independent      32.0   
1     Write-In - Evan McMullin          NaN       3.0   
2           Cassandra Mitchell  Independent      36.0   
3                          NaN          NaN       NaN   
4                 Gary Johnson  Independent       5.0   
...                        ...          ...       ...   
9397              Gary Johnson  Independent      47.0   
9398  Write-In - Evan McMullin  Independent      14.0   
9399                       NaN    

In [72]:
tn_2016.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9402 entries, 0 to 9401
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      9402 non-null   object 
 1   PRECINCT    9402 non-null   object 
 2   OFFICENAME  9402 non-null   object 
 3   ELECTDATE   9402 non-null   object 
 4   RNAME1      9402 non-null   object 
 5   PARTY1      7561 non-null   object 
 6   PVTALLY1    9402 non-null   int64  
 7   RNAME2      7984 non-null   object 
 8   PARTY2      6088 non-null   object 
 9   PVTALLY2    7984 non-null   float64
 10  RNAME3      5738 non-null   object 
 11  PARTY3      3787 non-null   object 
 12  PVTALLY3    5738 non-null   float64
 13  RNAME4      4913 non-null   object 
 14  PARTY4      2726 non-null   object 
 15  PVTALLY4    4913 non-null   float64
 16  RNAME5      4567 non-null   object 
 17  PARTY5      2726 non-null   object 
 18  PVTALLY5    4567 non-null   float64
 19  RNAME6      2269 non-null  

In [89]:
# merge entries in all RNAME columns into new column "name"
tn_2016_namemelt=tn_2016.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['RNAME1','RNAME2','RNAME3','RNAME4','RNAME5','RNAME6','RNAME7','RNAME8','RNAME9','RNAME10'],value_name="name")
# drop column "variable" created by .melt
tn_2016_namemelt=tn_2016_namemelt.drop(['variable'],axis=1)                          
tn_2016_namemelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson     Briceville   
...        ...            ...   
9400    Wilson       Absentee   
9401    Wilson       Absentee   
9402  Anderson  Andersonville   
9403  Anderson  Andersonville   
9404  Anderson  Andersonville   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2016-11-08   
1                               United States President  2016-11-08   
2     United States House of Representatives District 3  2016-11-08   
3        Tennessee House of Representatives District 36  2016-11-08   
4                               United States President  2016-11-08   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 46  2016-11-08   
9401     Tennessee House of Representatives District 57  2016-11-08   
9402                            United States President  2016-11-08   
9403                            United States President  2016-11-08   
9404  United States House of Representatives District 3  2016-11-08   

                                name  
0                    Donald J. Trump  
1     Write-In - Kyle Kenley Kopitke  
2                  Chuck Fleischmann  
3                   Dennis H. Powers  
4                    Donald J. Trump  
...                              ...  
9400                       Mark Pody  
9401                    Susan M Lynn  
9402                 Hillary Clinton  
9403   Write-In - Laurence Kotlikoff  
9404                  Melody Shekari  

[9405 rows x 5 columns]

In [91]:
# merge entries in all PARTY columns into new column "party"
tn_2016_partymelt=tn_2016.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PARTY1','PARTY2','PARTY3','PARTY4','PARTY5','PARTY6','PARTY7','PARTY8','PARTY9','PARTY10'],value_name="party")
# drop column "variable" created by .melt
tn_2016_partymelt=tn_2016_partymelt.drop(['variable'],axis=1)
tn_2016_partymelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson     Briceville   
...        ...            ...   
9400    Wilson       Absentee   
9401    Wilson       Absentee   
9402  Anderson  Andersonville   
9403  Anderson  Andersonville   
9404  Anderson  Andersonville   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2016-11-08   
1                               United States President  2016-11-08   
2     United States House of Representatives District 3  2016-11-08   
3        Tennessee House of Representatives District 36  2016-11-08   
4                               United States President  2016-11-08   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 46  2016-11-08   
9401     Tennessee House of Representatives District 57  2016-11-08   
9402                            United States President  2016-11-08   
9403                            United States President  2016-11-08   
9404  United States House of Representatives District 3  2016-11-08   

           party  
0     Republican  
1            NaN  
2     Republican  
3     Republican  
4     Republican  
...          ...  
9400  Republican  
9401  Republican  
9402  Democratic  
9403         NaN  
9404  Democratic  

[9405 rows x 5 columns]

In [92]:
# merge entries in all PVTALLY columns into new column "vote_count"
tn_2016_pvtallymelt=tn_2016.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PVTALLY1','PVTALLY2','PVTALLY3','PVTALLY4','PVTALLY5','PVTALLY6','PVTALLY7','PVTALLY8','PVTALLY9','PVTALLY10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2016_pvtallymelt=tn_2016_pvtallymelt.drop(['variable'],axis=1)
tn_2016_pvtallymelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson     Briceville   
...        ...            ...   
9400    Wilson       Absentee   
9401    Wilson       Absentee   
9402  Anderson  Andersonville   
9403  Anderson  Andersonville   
9404  Anderson  Andersonville   

                                             OFFICENAME   ELECTDATE  \
0                               United States President  2016-11-08   
1                               United States President  2016-11-08   
2     United States House of Representatives District 3  2016-11-08   
3        Tennessee House of Representatives District 36  2016-11-08   
4                               United States President  2016-11-08   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 46  2016-11-08   
9401     Tennessee House of Representatives District 57  2016-11-08   
9402                            United States President  2016-11-08   
9403                            United States President  2016-11-08   
9404  United States House of Representatives District 3  2016-11-08   

      vote_count  
0         1222.0  
1            0.0  
2         1124.0  
3         1128.0  
4          243.0  
...          ...  
9400       264.0  
9401       432.0  
9402       227.0  
9403         0.0  
9404       195.0  

[9405 rows x 5 columns]

In [93]:
# add "party" column to tn_2016_namemelt
party=tn_2016_partymelt['party']
tn_2016_namemelt=tn_2016_namemelt.join(party)
tn_2016_namemelt.head(3)

COUNTY       PRECINCT                                         OFFICENAME  \
0  Anderson  Andersonville                            United States President   
1  Anderson  Andersonville                            United States President   
2  Anderson  Andersonville  United States House of Representatives District 3   

    ELECTDATE                            name       party  
0  2016-11-08                 Donald J. Trump  Republican  
1  2016-11-08  Write-In - Kyle Kenley Kopitke         NaN  
2  2016-11-08               Chuck Fleischmann  Republican

In [95]:
# add "vote_count" column to tn_2016_namemelt
pvtally=tn_2016_pvtallymelt['vote_count']
tn_2016_namemelt=tn_2016_namemelt.join(pvtally)
tn_2016_namemelt.head(3)

COUNTY       PRECINCT                                         OFFICENAME  \
0  Anderson  Andersonville                            United States President   
1  Anderson  Andersonville                            United States President   
2  Anderson  Andersonville  United States House of Representatives District 3   

    ELECTDATE                            name       party  vote_count  
0  2016-11-08                 Donald J. Trump  Republican      1222.0  
1  2016-11-08  Write-In - Kyle Kenley Kopitke         NaN         0.0  
2  2016-11-08               Chuck Fleischmann  Republican      1124.0

In [96]:
tn_2016_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94020 entries, 0 to 94019
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      94020 non-null  object 
 1   PRECINCT    94020 non-null  object 
 2   OFFICENAME  94020 non-null  object 
 3   ELECTDATE   94020 non-null  object 
 4   name        43208 non-null  object 
 5   party       27969 non-null  object 
 6   vote_count  43208 non-null  float64
dtypes: float64(1), object(6)
memory usage: 5.0+ MB


In [98]:
# lower case column names
tn_2016_namemelt.columns=tn_2016_namemelt.columns.str.lower()
tn_2016_namemelt.head(2)

county       precinct               officename   electdate  \
0  Anderson  Andersonville  United States President  2016-11-08   
1  Anderson  Andersonville  United States President  2016-11-08   

                             name       party  vote_count  
0                 Donald J. Trump  Republican      1222.0  
1  Write-In - Kyle Kenley Kopitke         NaN         0.0

In [99]:
# change null values in "party" column to "no party"
tn_2016_namemelt['party'].fillna("No party", inplace = True)
tn_2016_namemelt.head(2)

county       precinct               officename   electdate  \
0  Anderson  Andersonville  United States President  2016-11-08   
1  Anderson  Andersonville  United States President  2016-11-08   

                             name       party  vote_count  
0                 Donald J. Trump  Republican      1222.0  
1  Write-In - Kyle Kenley Kopitke    No party         0.0

In [129]:
tn_2016_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94020 entries, 0 to 94019
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      94020 non-null  object 
 1   precinct    94020 non-null  object 
 2   officename  94020 non-null  object 
 3   electdate   94020 non-null  object 
 4   name        43208 non-null  object 
 5   party       94020 non-null  object 
 6   vote_count  43208 non-null  float64
dtypes: float64(1), object(6)
memory usage: 5.0+ MB


In [130]:
# change null values in "vote_count" column to "0"
tn_2016_namemelt['vote_count'].fillna(0)

0        1222.0
1           0.0
2        1124.0
3        1128.0
4         243.0
          ...  
94015       0.0
94016       0.0
94017       0.0
94018       0.0
94019       0.0
Name: vote_count, Length: 94020, dtype: float64

In [131]:
tn_2016_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94020 entries, 0 to 94019
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      94020 non-null  object 
 1   precinct    94020 non-null  object 
 2   officename  94020 non-null  object 
 3   electdate   94020 non-null  object 
 4   name        43208 non-null  object 
 5   party       94020 non-null  object 
 6   vote_count  43208 non-null  float64
dtypes: float64(1), object(6)
memory usage: 5.0+ MB


### 43208 non-null names and 43208 non-null vote counts
### Do the nulls in "name" correlate one to one with "no party"?

In [132]:
tn_2016_namemelt[tn_2016_namemelt['name'].isna()]

county             precinct  \
9573   Benton           1 Holladay   
9578   Benton     2 Camden Jr High   
9583   Benton   3 Camden City Hall   
9588   Benton  4 Camden Elementary   
9593   Benton            5 Library   
...       ...                  ...   
94014  Wilson          Provisional   
94016  Wilson             Absentee   
94017  Wilson             Absentee   
94018  Wilson             Absentee   
94019  Wilson             Absentee   

                                              officename   electdate name  \
9573                        Tennessee Senate District 24  2016-11-08  NaN   
9578                        Tennessee Senate District 24  2016-11-08  NaN   
9583                        Tennessee Senate District 24  2016-11-08  NaN   
9588                        Tennessee Senate District 24  2016-11-08  NaN   
9593                        Tennessee Senate District 24  2016-11-08  NaN   
...                                                  ...         ...  ...   
94014     Tennessee House of Representatives District 57  2016-11-08  NaN   
94016                            United States President  2016-11-08  NaN   
94017  United States House of Representatives District 6  2016-11-08  NaN   
94018     Tennessee House of Representatives District 46  2016-11-08  NaN   
94019     Tennessee House of Representatives District 57  2016-11-08  NaN   

          party  vote_count  
9573   No party         NaN  
9578   No party         NaN  
9583   No party         NaN  
9588   No party         NaN  
9593   No party         NaN  
...         ...         ...  
94014  No party         NaN  
94016  No party         NaN  
94017  No party         NaN  
94018  No party         NaN  
94019  No party         NaN  

[50812 rows x 7 columns]

In [151]:
tn_2016_namemelt[tn_2016_namemelt['name'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50812 entries, 9573 to 94019
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      50812 non-null  object 
 1   precinct    50812 non-null  object 
 2   officename  50812 non-null  object 
 3   electdate   50812 non-null  object 
 4   name        0 non-null      object 
 5   party       50812 non-null  object 
 6   vote_count  0 non-null      float64
dtypes: float64(1), object(6)
memory usage: 3.1+ MB


### Looks safe to drop rows with no votes tallied

In [155]:
# drop rows with no name and vote count
tn_2016_final=tn_2016_namemelt.dropna()

In [156]:
tn_2016_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43208 entries, 0 to 94015
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      43208 non-null  object 
 1   precinct    43208 non-null  object 
 2   officename  43208 non-null  object 
 3   electdate   43208 non-null  object 
 4   name        43208 non-null  object 
 5   party       43208 non-null  object 
 6   vote_count  43208 non-null  float64
dtypes: float64(1), object(6)
memory usage: 2.6+ MB


In [157]:
tn_2016_final=tn_2016_final.astype({'vote_count':'int64'}).astype({'vote_count':'int64'})

In [158]:
tn_2016_final

county       precinct  \
0      Anderson  Andersonville   
1      Anderson  Andersonville   
2      Anderson  Andersonville   
3      Anderson  Andersonville   
4      Anderson     Briceville   
...         ...            ...   
93997    Wilson           24-1   
94001    Wilson           25-1   
94005    Wilson           25-2   
94010    Wilson    Provisional   
94015    Wilson       Absentee   

                                              officename   electdate  \
0                                United States President  2016-11-08   
1                                United States President  2016-11-08   
2      United States House of Representatives District 3  2016-11-08   
3         Tennessee House of Representatives District 36  2016-11-08   
4                                United States President  2016-11-08   
...                                                  ...         ...   
93997                            United States President  2016-11-08   
94001                            United States President  2016-11-08   
94005                            United States President  2016-11-08   
94010                            United States President  2016-11-08   
94015                            United States President  2016-11-08   

                                 name        party  vote_count  
0                     Donald J. Trump   Republican        1222  
1      Write-In - Kyle Kenley Kopitke     No party           0  
2                   Chuck Fleischmann   Republican        1124  
3                    Dennis H. Powers   Republican        1128  
4                     Donald J. Trump   Republican         243  
...                               ...          ...         ...  
93997         Write-In - Tom Hoefling  Independent           0  
94001         Write-In - Tom Hoefling  Independent           0  
94005         Write-In - Tom Hoefling  Independent           0  
94010         Write-In - Tom Hoefling  Independent           0  
94015         Write-In - Tom Hoefling  Independent           0  

[43208 rows x 7 columns]

In [159]:
tn_2016_final['officename'].value_counts()

United States President                              30330
United States House of Representatives District 8     1729
United States House of Representatives District 3     1380
United States House of Representatives District 1      868
United States House of Representatives District 7      852
                                                     ...  
Tennessee House of Representatives District 99          12
Tennessee House of Representatives District 1           11
Tennessee House of Representatives District 68          11
Tennessee House of Representatives District 16          10
Tennessee House of Representatives District 24          10
Name: officename, Length: 125, dtype: int64

In [163]:
tn_2016_final['officename'].unique()

array(['United States President',
       'United States House of Representatives District 3',
       'Tennessee House of Representatives District 36',
       'Tennessee House of Representatives District 33',
       'United States House of Representatives District 4',
       'Tennessee Senate District 14',
       'Tennessee House of Representatives District 62',
       'United States House of Representatives District 7',
       'Tennessee Senate District 24',
       'Tennessee House of Representatives District 75',
       'United States House of Representatives District 8',
       'Tennessee House of Representatives District 31',
       'United States House of Representatives District 2',
       'Tennessee Senate District 2',
       'Tennessee House of Representatives District 8',
       'Tennessee House of Representatives District 20',
       'Tennessee Senate District 10',
       'Tennessee House of Representatives District 24',
       'Tennessee House of Representatives District 22',